armi/physics/neutronics/fissionProductModel/fissionProductModel.py

```diff
@@ -205,60 +192,146 @@ def _getAllFissionProductNames(self):
-        The CrossSectionGroupManager does this for each XSG individually for
-        lattice physics, but it's important to keep these up to date as well for
-        anything else that may be interested in fission product information
-        (e.g. MCNP).
-
-        See Also
-        --------
-        armi.physics.neutronics.crossSectionGroupManager.AverageBlockCollection.getRepresentativeBlock
-
+        The current implementation will update the number density vector
+        of each of the blocks and the gaseous fission products are not
+        moved or displaced to another area in the core.
         """
-        runLog.extra("Updating lumped fission product gas removal fractions")
-        avgGasReleased = 0.0
-        totalWeight = 0.0
-        for block in self.r.core.getBlocks(Flags.FUEL):
-            if self._useGlobalLFPs:
-                # add to average for globals
-                weight = block.getVolume() * (block.p.flux or 1.0)
-                avgGasReleased += block.p.gasReleaseFraction * weight
-                totalWeight += weight
-            else:
-                # set individually
-                block.getLumpedFissionProductCollection().setGasRemovedFrac(
-                    block.p.gasReleaseFraction
+        if not self.cs["fgRemoval"]:
+            runLog.info(
+                "Skipping removal of gaseous fission products since the `fgRemoval` setting is disabled."
+            )
+            return
+
+        runLog.info(f"Removing the gaseous fission products from the core.")
+        if not isinstance(gasRemovalFractions, dict):
+            raise TypeError(f"The gas removal fractions input is not a dictionary.")
+
+        # Check that the gas removal fractions supplied are within [0.0, 1.0] and cap them
+        # at these limits otherwise.
+        updatedGasRemovalFractions = {}
+        for b, frac in gasRemovalFractions.items():
+            if frac < 0.0:
+                runLog.warning(
+                    f"The fission gas removal fraction is less than zero for {b}. Setting to zero."
+                )
+                updatedGasRemovalFractions[b] = 0.0
+            elif frac > 1.0:
+                runLog.warning(
+                    f"The fission gas removal fraction is greater than one for {b}. Setting to one."
                 )
+                updatedGasRemovalFractions[b] = 1.0
+            else:
+                updatedGasRemovalFractions[b] = frac
+        gasRemovalFractions.update(updatedGasRemovalFractions)
+
+        def _getGaseousFissionProductNumberDensities(b, lfp):
+            """Look into a single lumped fission product object and pull out the gaseous atom number densities."""
+            numberDensities = {}
+            for nb in lfp.keys():
+                if not lumpedFissionProduct.isGas(nb):
+                    continue
+                yld = lfp[nb]
+                ndens = b.getNumberDensity(lfp.name)
+                numberDensities[nb.name] = ndens * yld
```

Comment:
I think this evaluation will only work as expected the before any gaseous fission products have been removed from the LFP.

For instance, you call this method from within `_removeFissionGasesLumpedFissionProductModeling()`. The first time you execute that method, this logic works fine. But once `_removeFissionGasesLumpedFissionProductModeling()` has actually modified the number density of the LFP, then yields are no longer representative of the isotopic split within the LFP. So if you call `_removeFissionGasesLumpedFissionProductModeling()` _again_, the value returned by this method will actually be over-representing the number of gaseous fission products in the LFP. Assuming that `_removeFissionGasesLumpedFissionProductModeling()` will be called on every time node, this will be a problem.

Does that make sense?

I think the only way around this is through some sort of caching from one time step to the next.

This is _maybe_ why the fission yields were being modified in the previous implementation, because that would account for the cumulative impact of the fission gases disappearing from the LFP.

Reply:
Fission yields are a generation term rather than a removal term. I am going to modify this implementation based on a conversation with @sammiller11235 and @mgjarrett. If it takes too long or isn't seemingly correct I might just remove this functionality altogether and then create an issue for determining the requirements. I think that there are enough "methodology" questions to be asked here that a generalized framework implementation might not cut it. Not sure yet.